### PR TITLE
Fix: Check if no allocation is necessary in jagged vec buffer

### PR DESCRIPTION
**Pull request by guilhermeAlmeida1**

### CODE REVIEW

**Review by krasznaa — CHANGES_REQUESTED** [review]
I don't think this is quite right... :thinking:

If somebody asks for a jagged vector whose "inner vectors" are all empty, this code change will end up not initialising the "outer memory" array(s) correctly. Since for such an object one still needs to create `vecmem::data::vector_view` objects, all with `0` sizes. (Otherwise code inspecting such an "empty" container would crash.)

Also, for a resizable buffer with all-zero capacities, we still need to allocate the "size variables" in memory. Even if they will only ever be allowed to take the value `0`. (Think for instance that for a jagged vector for which only a couple of inner vectors have a `0` capacity, we also still create the "size variables" for those inner vectors.)

Generally, please add some unit tests that explicitly test partially, and fully empty jagged vector buffers. Both resizable and non-resizable versions of them. Then we will be sure that this PR would work correctly in all circumstances. :wink:

**Comment by guilhermeAlmeida1** [comment]
I'm still trying to figure out how to make this work properly, but at least #232 seems to have made this work properly without any errors, and all the tests I've added pass without changing anything in the source code (I haven't locally tested hip yet but I don't expect to see any different behaviour there).

**Review by krasznaa — CHANGES_REQUESTED** [review]
I like the direction, but I think it's more complicated right now than it needs to be.

File: core/include/vecmem/containers/impl/jagged_vector_buffer.ipp [hunk 1]
```diff
@@ -116,16 +126,25 @@ jagged_vector_buffer<TYPE>::jagged_vector_buffer(
...
         } else {
-            new (base_type::host_ptr() + i) value_type(
-                static_cast<typename value_type::size_type>(capacities[i]),
-                data_ptr + ptrdiff);
+            if (header_ptr != nullptr) {
+                new (base_type::host_ptr() + i)
+                    value_type(0, &(header_ptr[i]), nullptr);
+            } else {
+                new (base_type::host_ptr() + i) value_type(0, nullptr);
+            }
```
[the remark this answers is not in the document]
Comment (guilhermeAlmeida1) [reply core/include/vecmem/containers/impl/jagged_vector_buffer.ipp:146]: It's not, like I mentioned here before, the whole code works fine without any changes since #232 was merged in. I can just leave in the 1st change of not allocating device memory for `fixed_size` buffers with 0 total size.
My point was to try to match the innards of the `jagged_vector_buffer` to what the `vector_buffer` does with using `nullptr` for emptyness, and also trying to reduce the ammount of unnecessary calculations being made in these cases, altough I guess this might not be achieving so much anyways.
